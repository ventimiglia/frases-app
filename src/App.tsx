import React, { useState } from "react";
import CardGrid from "./components/CardGrid";
import SearchBar from "./components/Forms/SearchBar";
import SavePhrase from "./components/Forms/SavePhrase"

const App: React.FC = () => {
  const [phrase, setPhrase] = useState<string>("");
  const [phrasesSaved, setPhrasesSaved] = useState<string[]>([]);
  const [filteredPhrases, setFilteredPhrases] = useState<string[]>([]);


  const handleSearch = (text: string) => {
    const filtered = phrasesSaved.filter((phrase) =>
      phrase.toLowerCase().includes(text.toLowerCase())
    );
    setFilteredPhrases(filtered);
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setPhrase(e.target.value);
  };

  const onSave = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    setPhrasesSaved([...phrasesSaved, phrase]);
    setFilteredPhrases([...filteredPhrases, phrase]);
    setPhrase("");
  };

  return (
    <main className="w-screen h-screen mx-auto">
      <header className="h-2/5  px-8 w-full bg-primary flex shadow-lg">
        <SavePhrase onSave={onSave} onChange={handleChange} phrase={phrase} />
      </header>
      <section className="text-zinc-800 h-3/5 p-8">
        <div className="h-full w-full border-[1px] border-gray-300 shadow-lg rounded-xl px-8 py-2 overflow-y-auto">
          {phrasesSaved.length > 0 && <SearchBar onSearch={handleSearch} />}
          <CardGrid phrases={filteredPhrases ?? phrasesSaved} />
        </div>
      </section>
    </main>
  );
};

export default App;
