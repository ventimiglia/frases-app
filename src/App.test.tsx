import { fireEvent, render } from "@testing-library/react";
import App from "./App";

describe("App", () => {
  it("should render without cards", () => {
    const { container, getByText, getByPlaceholderText, getByLabelText } =
    render(<App />);
    const input = getByPlaceholderText("Ahora un poema...");
    const label = getByLabelText("Agregar frase");
    const button = getByText("Guardar");
    
    expect(input).toBeInTheDocument();
    expect(label).toBeInTheDocument();
    expect(button).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("should save a phrase and display it in the grid", () => {
    const { container, getByText, getByPlaceholderText, getByRole } =
    render(<App />);
    const input = getByPlaceholderText("Ahora un poema...");
    const button = getByText("Guardar");
    
    fireEvent.change(input, { target: { value: "Test phrase" } });
    fireEvent.click(button);

    const card = getByRole("article");

    expect(card).toBeInTheDocument();
    expect(card).toHaveTextContent("Test phrase");
    expect(container).toMatchSnapshot();
  })

  it("should filter the grid by the search input", () => {
    const { getByText, getByPlaceholderText, getByLabelText } =
    render(<App />);
    const input = getByPlaceholderText("Ahora un poema...");
    const button = getByText("Guardar");
    const phrases = ["Phrase", "Poem", "Letter"];

    for(let phrase of phrases) {
      fireEvent.change(input, { target: { value: phrase } });
      fireEvent.click(button);
    }

    const phrase1 = getByText("Phrase");
    const phrase2 = getByText("Poem");
    const phrase3 = getByText("Letter");
    
    expect(phrase1).toBeInTheDocument();
    expect(phrase2).toBeInTheDocument();
    expect(phrase3).toBeInTheDocument();
    
    const searchInput = getByLabelText("Buscar frase");
    fireEvent.change(searchInput, { target: { value: "Phrase" } });
    
    expect(phrase1).toBeInTheDocument();
    expect(phrase2).not.toBeInTheDocument();
    expect(phrase3).not.toBeInTheDocument();
  })
});
