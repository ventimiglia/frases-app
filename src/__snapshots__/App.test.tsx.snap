// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render without cards 1`] = `
<div>
  <main
    class="w-screen h-screen mx-auto"
  >
    <header
      class="h-2/5  px-8 w-full bg-primary flex shadow-lg"
    >
      <form
        class="flex flex-col gap-2 w-full self-center max-w-4xl m-auto"
      >
        <label
          class="flex flex-col h-full w-4/5 text-2xl font-bold"
          for="phrase"
        >
          Agregar frase
        </label>
        <div
          class="flex md:h-14 shadow-lg flex-col md:flex-row gap-4 md:gap-0"
        >
          <input
            class="w-full md:w-4/5 text-primary text-3xl px-4 focus:outline-none md:rounded-s-md"
            id="phrase"
            name="phrase"
            placeholder="Ahora un poema..."
            value=""
          />
          <button
            class="w-full md:w-1/5 text-3xl border-2 border-white transition-colors hover:bg-purple-600 md:rounded-e-md"
            type="submit"
          >
            Guardar
          </button>
        </div>
      </form>
    </header>
    <section
      class="text-zinc-800 h-3/5 p-8"
    >
      <div
        class="h-full w-full border-[1px] border-gray-300 shadow-lg rounded-xl px-8 py-2 overflow-y-auto"
      >
        <div
          class="grid grid-cols-3 gap-4"
        />
      </div>
    </section>
  </main>
</div>
`;

exports[`App should save a phrase and display it in the grid 1`] = `
<div>
  <main
    class="w-screen h-screen mx-auto"
  >
    <header
      class="h-2/5  px-8 w-full bg-primary flex shadow-lg"
    >
      <form
        class="flex flex-col gap-2 w-full self-center max-w-4xl m-auto"
      >
        <label
          class="flex flex-col h-full w-4/5 text-2xl font-bold"
          for="phrase"
        >
          Agregar frase
        </label>
        <div
          class="flex md:h-14 shadow-lg flex-col md:flex-row gap-4 md:gap-0"
        >
          <input
            class="w-full md:w-4/5 text-primary text-3xl px-4 focus:outline-none md:rounded-s-md"
            id="phrase"
            name="phrase"
            placeholder="Ahora un poema..."
            value=""
          />
          <button
            class="w-full md:w-1/5 text-3xl border-2 border-white transition-colors hover:bg-purple-600 md:rounded-e-md"
            type="submit"
          >
            Guardar
          </button>
        </div>
      </form>
    </header>
    <section
      class="text-zinc-800 h-3/5 p-8"
    >
      <div
        class="h-full w-full border-[1px] border-gray-300 shadow-lg rounded-xl px-8 py-2 overflow-y-auto"
      >
        <label
          class="flex flex-col gap-1 mb-4 animate-fade-right animate-duration-500 text-lg"
          for="search"
        >
          Buscar frase
          <input
            class="w-full border-b-[1px] focus:outline-none text-lg"
            id="search"
            name="search"
            placeholder="Ahora un poema ..."
            type="text"
          />
        </label>
        <div
          class="grid grid-cols-3 gap-4"
        >
          <article
            class="flex justify-center items-center text-lg border-[1px] border-secondary/60 h-24 rounded-md shadow-lg animate-fade-up animate-duration-500 p-2 overflow-auto"
          >
            <p>
              Test phrase
            </p>
          </article>
        </div>
      </div>
    </section>
  </main>
</div>
`;
